LLama/Model/ModelManager.cs

```diff
@@ -0,0 +1,309 @@
+                {
+                    FileName = fi.Name,
+                    FilePath = fi.FullName,
+                    ModelType = ModelFileType.GGUF,
+                    SizeInBytes = fi.Length,
+                });
+            }
+
+            _availableModels.Add(fullDirectoryPath, directoryModelFiles);
+        }
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ModelFileMetadata> ModelFileList
+        => _availableModels.SelectMany(x => x.Value);
+    /// <inheritdoc />
+    public IEnumerable<string> ModelDirectories
+        => _availableModels.Keys;
+
+    /// <inheritdoc />
+    public void AddDirectory(string directory)
+    {
+        GetModelsFromDirectories(directory);
+    }
+
+    /// <inheritdoc />
+    public bool RemoveDirectory(string directory)
+    {
+        return _availableModels.Remove(Path.GetFullPath(directory));
+    }
+
+    /// <inheritdoc />
+    public void RemoveAllDirectories()
+    {
+        _availableModels.Clear();
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ModelFileMetadata> GetAvailableModelsFromDirectory(string directory)
+    {
+        var dirPath = Path.GetFullPath(directory);
+        return _availableModels.TryGetValue(dirPath, out var dirModels)
+            ? dirModels
+            : [];
+    }
+
+    /// <inheritdoc />
+    public bool TryGetModelFileMetadata(string fileName, out ModelFileMetadata modelMeta)
+    {
+        var filePath = Path.GetFullPath(fileName);
+        modelMeta = ModelFileList.FirstOrDefault(f => f.FilePath == filePath)!;
+        return modelMeta != null;
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<LLamaWeights> GetLoadedModels()
+    {
+        return _loadedModelCache.Values;
+    }
+
+    /// <inheritdoc />
+    public bool TryGetLoadedModel(string modelId, out LLamaWeights model)
+    {
+        return _loadedModelCache.TryGetValue(modelId, out model!);
+    }
+
+    /// <inheritdoc />
+    public async Task<LLamaWeights> LoadModel(string modelPath,
+        Action<ModelParams>? modelConfigurator = null!,
+        string modelId = "",
+        CancellationToken cancellationToken = default)
+    {
+        // Configure model params
+        var modelParams = new ModelParams(modelPath);
+        modelConfigurator ??= DefaultModelConfigurator;
+        modelConfigurator.Invoke(modelParams);
+
+        // load and cache
+        var model = await LLamaWeights.LoadFromFileAsync(modelParams, cancellationToken);
+        if (string.IsNullOrWhiteSpace(modelId))
+        {
+            modelId = model.ModelName;
+        }
+        _loadedModelCache.Add(modelId, model);
+        return model;
+    }
+
+    /// <summary>
+    /// Updates the passed in model params with default values
+    /// </summary>
+    /// <param name="modelParams"></param>
+    public static void DefaultModelConfigurator(ModelParams modelParams)
+    {
+        // Reasonable defaults
+        modelParams.GpuLayerCount = 12;
```

Comment:
`ModelParams` already has default values for all properties which are meant to be reasonable defaults. I think it'd be better to use those values.